process/process.go

```diff
@@ -0,0 +1,380 @@
+
+		proposer:    proposer,
+		validator:   validator,
+		observer:    observer,
+		broadcaster: broadcaster,
+		scheduler:   scheduler,
+		timer:       timer,
+	}
+	p.StartRound(0)
+	return p
+}
+
+// MarshalJSON implements the `json.Marshaler` interface for the Process type,
+// by marshaling its isolated State.
+func (p Process) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.state)
+}
+
+// UnmarshalJSON implements the `json.Unmarshaler` interface for the Process
+// type, by unmarshaling its isolated State.
+func (p *Process) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, &p.state)
+}
+
+// StartRound is safe for concurrent use. See
+// https://arxiv.org/pdf/1807.04938.pdf for more information.
+func (p *Process) StartRound(round block.Round) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.startRound(round)
+}
+
+// HandleMessage is safe for concurrent use. See
+// https://arxiv.org/pdf/1807.04938.pdf for more information.
+func (p *Process) HandleMessage(m Message) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	switch m := m.(type) {
+	case *Propose:
+		p.handlePropose(m)
+	case *Prevote:
+		p.handlePrevote(m)
+	case *Precommit:
+		p.handlePrecommit(m)
+	}
+}
+
+func (p *Process) startRound(round block.Round) {
+	p.state.CurrentRound = round
+	p.state.CurrentStep = StepPropose
+	if p.signatory.Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+		var proposal block.Block
+		if p.state.ValidBlock.Hash() != block.InvalidHash {
+			proposal = p.state.ValidBlock
+		} else {
+			proposal = p.proposer.BlockProposal(p.state.CurrentHeight, p.state.CurrentRound)
+		}
+		p.broadcaster.Broadcast(NewPropose(
+			p.state.CurrentHeight,
+			p.state.CurrentRound,
+			proposal,
+			p.state.ValidRound,
+		))
+	} else {
+		p.scheduleTimeoutPropose(p.state.CurrentHeight, p.state.CurrentRound, p.timer.Timeout(StepPropose, p.state.CurrentRound))
+	}
+}
+
+func (p *Process) handlePropose(propose *Propose) {
+	_, firstTime, _, _ := p.state.Proposals.Insert(propose)
+
+	// upon Propose{currentHeight, currentRound, block, -1}
+	if propose.Height() == p.state.CurrentHeight && propose.Round() == p.state.CurrentRound && propose.ValidRound() == block.InvalidRound {
+		// from Schedule{currentHeight, currentRound}
+		if propose.Signatory().Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+			// while step = StepPropose
+			if p.state.CurrentStep == StepPropose {
+				if p.validator.IsBlockValid(propose.Block()) && (p.state.LockedRound == block.InvalidRound || p.state.LockedBlock.Equal(propose.Block())) {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						propose.Block().Hash(),
+					))
+				} else {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						block.InvalidHash,
+					))
+				}
+				p.state.CurrentStep = StepPrevote
+			}
+		}
+	}
+
+	p.checkProposeInCurrentHeightAndRoundWithPrevotes()
```

Comment:
This makes more sense to be moved into the above `if` struct (after modifying it as follows):
```
// while step = StepPropose
if p.state.CurrentStep == StepPropose && propose.ValidRound() == block.InvalidRound{
    .
    .
    .
} else {
    p.checkProposeInCurrentHeightAndRoundWithPrevotes()
}
```

Reply:
It is possible that both conditions are true. The Tendermint algorithm description does not mutually exclude the two conditions, so I don't think putting it in an else really makes sense. As far as I understand, it is completely possible that `checkProposeInCurrentHeightAndRoundWithPrevotes` is needed when `p.state.CurrentStep == StepPropose && propose.ValidRound() == block.InvalidRound` is true.

Reply:
Yes, what I meant was that if `p.state.CurrentStep == StepPropose && propose.ValidRound() == block.InvalidRound` is true (i.e., line 22 of the algorithm is true), then `p.state.CurrentStep = StepPrevote` (line 27) will be evaluated, which in turn means that `checkProposeInCurrentHeightAndRoundWithPrevotes` would not do anything useful because it checks if `p.state.CurrentStep == StepPropose` inorder to proceed (line 28). 

Again, this doesn't really make a big difference, so happy to leave it as is.

Reply:
@divyakoshy true, I can see that argument. However, I would prefer to keep the direct comparison between the implementation and the formally described algorithm. Keeping it in its current form means you don't have to couple the internal implementation details of these functions (which break isolation).